### PR TITLE
feat(profiling): Run profiling on self-hosted

### PR DESCRIPTION
**Pull request by phacops**
This PR aims to add support for profiling on self-hosted.

Fixes https://github.com/getsentry/self-hosted/issues/1838.

### CODE REVIEW

File: docker-compose.yml
Comment (aldy505) [inline docker-compose.yml:389]:
```suggestion
    image: "$VROOM_IMAGE"
```

Should be inferred from variable

**Review by aldy505 — COMMENTED** [review]
I don't see any configuration for integrating it with the `web` or `snuba` containers? Or is it meant to be just like this? I'd like to try it on my own server 😄

**Comment by phacops** [comment]
@aldy505 Good point, I added those.

File: docker-compose.yml
Comment (aldy505) [inline docker-compose.yml:424]: Don't forget to add a line about this here: https://github.com/getsentry/self-hosted/blob/master/install/create-docker-volumes.sh

File: docker-compose.yml
Comment (aldy505) [inline docker-compose.yml:291]: Should these be added to the list of sentry image's depends_on field?

Here: https://github.com/getsentry/self-hosted/blob/8aa571cb758406eccb7aefe13b3b7ebf1a34e5c2/docker-compose.yml#L39-L50

File: docker-compose.yml
Comment (aldy505) [inline docker-compose.yml:403]: I don't think this reads from environment variable defined here, as this is hardcoded.

https://github.com/getsentry/self-hosted/blob/8aa571cb758406eccb7aefe13b3b7ebf1a34e5c2/docker-compose.yml#L60

Should we also hardcode the address of Snuba API host?

Unless you want to do `"${SNUBA:-http://snuba-api:3128}"`

File: docker-compose.yml
Comment (aldy505) [inline docker-compose.yml:396]: Shouldn't we add the `vroom` address to the `x-sentry-default` key (or `web` container) as defined here?

https://github.com/getsentry/sentry/blob/39233a25d1bee5fd4dffa754b0b618128ef3deaa/src/sentry/conf/server.py#L3201

**Review by aldy505 — COMMENTED** [review]
Actually I'm a bit curious about the `SENTRY_DSN` environment variable that can be applied to snuba and vroom container. If they're left empty, I suppose no error message will be sent to the `internal` project of people's self-hosted  instance.

Vroom's defined here: https://github.com/getsentry/vroom/blob/c2cf9653f5bb9b2e4a02f0548956b790547e9048/cmd/vroom/config.go#L8

File: docker-compose.yml
Comment (hubertdeng123) [inline docker-compose.yml:360]: small thing, can this be moved up to be under these lines?

https://github.com/getsentry/self-hosted/blob/8aa571cb758406eccb7aefe13b3b7ebf1a34e5c2/docker-compose.yml#L332-L337

This file is pretty large and it would keep things a bit more organized

File: docker-compose.yml
Comment (hubertdeng123) [reply docker-compose.yml:403]: +1, seeing this in CI

https://github.com/getsentry/self-hosted/actions/runs/5015430579/jobs/9018870128?pr=2154

``` shell
+ docker compose --ansi never run --rm web createuser --force-update --superuser --email test@example.com -***
  time="2023-05-19T21:14:23Z" level=warning msg="The \"SNUBA\" variable is not set. Defaulting to a blank string."
```

**Review by hubertdeng123 — COMMENTED** [review]
Will profiling be released under CalVer? If so, would need to add `VROOM` to 
https://github.com/getsentry/self-hosted/blob/1fa77341a49180755edf84450419f9733fce0bab/scripts/bump-version.sh#L10

**Comment by phacops** [comment]
> Actually I'm a bit curious about the `SENTRY_DSN` environment variable that can be applied to snuba and vroom container. If they're left empty, I suppose no error message will be sent to the `internal` project of people's self-hosted instance.

That's correct, the Sentry SDK won't be activated and nothing will be sent. I was expecting the user to pass the right DSN via an environment variable. Do you know how other services are handling that?

**Comment by aldy505** [comment]
> That's correct, the Sentry SDK won't be activated and nothing will be sent. I was expecting the user to pass the right DSN via an environment variable. Do you know how other services are handling that?

We might as well need to ping @hubertdeng123 for help. 

There is an "internal" project that should've been set up during the first install of Sentry, but I don't know on how to acquire the DSN and put it into the environment variable / configuration that can be consumed by Snuba and Vroom.

**Comment by hubertdeng123** [comment]
> There is an "internal" project that should've been set up during the first install of Sentry, but I don't know on how to acquire the DSN and put it into the environment variable / configuration that can be consumed by Snuba and Vroom.

Snuba is very similar, [retrieves `SENTRY_DSN` from the environment](https://github.com/getsentry/snuba/blob/9d503a951019a4ebad5df6d418c9a950ca6c8595/snuba/settings/settings_self_hosted.py#L25). I've never played around with it before but it should just be possible to put it into the configuration through the `docker-compose.yml` file. I don't think this is a large concern for this PR though. For the record, looks like a DSN for the internal project can be shown by going to `/settings/${org}/projects/internal/keys/`

**Comment by aldy505** [comment]
I guess we can skip that one for now?

**Comment by hubertdeng123** [comment]
> I guess we can skip that one for now?

Yep

File: _unit-test/create-docker-volumes-test.sh
Comment (hubertdeng123) [inline _unit-test/create-docker-volumes-test.sh:24]: not sure why this was ever duplicated even after looking through the file history, and tests seem to be passing fine without it

File: .env
Comment (aldy505) [inline .env:12]: wait.. whaaaat? 🫨

File: _unit-test/create-docker-volumes-test.sh
Comment (aldy505) [reply _unit-test/create-docker-volumes-test.sh:24]: maybe testing for duplicate volume creation?

File: .env
Comment (phacops) [reply .env:12]: This should be `nightly`, for testing purposes i changed it but now we are finally publishing `nightly` tags properly.

Comment (aldy505) [reply .env]: Ah ok. Understandable.

File: .env
Comment (BYK) [inline .env:17]: Why is this an env variable and not in a config file?

File: _unit-test/create-docker-volumes-test.sh
Comment (BYK) [inline _unit-test/create-docker-volumes-test.sh:18]: I feel like this is not a great practice (creating and relying on global volumes). Can we not make this a "local" volume? Do we need to absolutely make sure the data here should survive a potentially accidental `docker compose down -v`?

File: docker-compose.yml
Comment (BYK) [inline docker-compose.yml:286]: Yay, more snuba containers! Let's add another strike to getsentry/snuba#1670

File: docker-compose.yml
Comment (BYK) [reply docker-compose.yml:395]: It should be in the config file below (or through the env variable as you suggested, yes). Otherwise it won't (shouldn't?) work

File: scripts/bump-version.sh
Comment (BYK) [inline scripts/bump-version.sh:10]: Maybe we should consider making this a blocklist now since almost all images get updated?

**Review by BYK — COMMENTED** [review]
There should be at least a cursory test sending a profile and making sure it is ingested before this is merged. Otherwise someone will eventually break and nobody will notice it until its too late :) (true story)

**Comment by BYK** [comment]
Regarding `SENTRY_DSN`, it _is_ created during the initial setup. It's just tricky to extract it out of the container and then putting in a "permanent" and shared place for other containers to consume. Can be dumped into a shared volume maybe?

File: .env
Comment (aldy505) [reply .env:17]: It's because.. uhhh... https://github.com/getsentry/vroom/blob/bdda64cc531de12ff86fc2a547850712a231ceeb/cmd/vroom/config.go#L4-L19

Comment (aldy505) [reply .env:17]: (Git blamed to my commit lol)

File: _unit-test/create-docker-volumes-test.sh
Comment (phacops) [reply _unit-test/create-docker-volumes-test.sh:18]: This will end up with profile data, it's something you'd likely want persisted by default.

**Comment by phacops** [comment]
> There should be at least a cursory test sending a profile and making sure it is ingested before this is merged. Otherwise someone will eventually break and nobody will notice it until its too late :) (true story)

It was added.

File: _unit-test/create-docker-volumes-test.sh
Comment (BYK) [reply _unit-test/create-docker-volumes-test.sh:18]: Local volumes are also persisted by default. They are just not "globally" available by docker containers outside of this project and you can remove them with `docker compose down -v`.

I'd say we should stop adding new global volumes and keep everything local unless we have _very_ good reasons and we understand them fully.

File: _unit-test/create-docker-volumes-test.sh
Comment (BYK) [reply _unit-test/create-docker-volumes-test.sh:24]: Yes, this was to make sure trying to create volumes over and over again does not error. I think, since we run the whole installation twice now, this is redundant.

File: fixtures/envelope-with-profile.json
Comment (BYK) [inline fixtures/envelope-with-profile.json:5]: Not saying this is too large but if there's a "minimally viable profile file" that's smaller than this, I'd prefer that 😅

File: _unit-test/create-docker-volumes-test.sh
Comment (aldy505) [reply _unit-test/create-docker-volumes-test.sh:18]: Profiling data should survive accidental `docker compose down`, though. Although AFAIK, [vroom](https://github.com/getsentry/vroom) doesn't have a cleanup mechanism for profiling data that exceeds the maximum sentry retention period. I guess for now, every profiling data stored on this volume will be persisted forever.

File: fixtures/envelope-with-profile.json
Comment (hubertdeng123) [inline fixtures/envelope-with-profile.json:1]: This doesn't seem to be a valid json file, seems to be an envelope file as far as I can tell? Maybe that's why the test is failing?

File: _unit-test/create-docker-volumes-test.sh
Comment (chadwhitacre) [reply _unit-test/create-docker-volumes-test.sh:18]: Switching from global to local volumes seems like a reasonable improvement, but I don't think we should make @phacops implement that for us. Let's let him use the status quo patterns and make the switch in a separate PR.

Comment (BYK) [reply _unit-test/create-docker-volumes-test.sh:18]: @chadwhitacre this one does not require a refactor. Just not adding it here and removing the `external: true` line in the compose file will do it. We already have other local volumes.

File: fixtures/envelope-with-profile.json
Comment (phacops) [reply fixtures/envelope-with-profile.json:1]: Test was failing because `--data` wasn't respecting new lines. Using `--data-binary` solves this.

File: _unit-test/create-docker-volumes-test.sh
Comment (phacops) [reply _unit-test/create-docker-volumes-test.sh:18]: @BYK It's definitely an issue we didn't really think about since we were relying on GCS retention settings to remove profiles when needed. I should be addressed separately in `vroom`.

Comment (phacops) [reply _unit-test/create-docker-volumes-test.sh:18]: @BYK Now, the reason why I set the `external: true` was because I wanted the volume to be persisted and according to the comment we have for non-external volumes, it suggests they won't be persisted. Seems like the comment is misguided though.

File: docker-compose.yml
Comment (aldy505) [reply docker-compose.yml:395]: We should be good to go after this one is addressed, right? @phacops

File: docker-compose.yml
Comment (BYK) [inline docker-compose.yml:427]: We should remove this comment as you suggested earlier to not confuse future folks. Definitely not accurate. (I mean we don't _need_ this data to persist through restarts but they do regardless)
```suggestion
```

**Review by BYK — APPROVED** [review]
🚀

File: .env
Comment (phacops) [reply .env:17]: @BYK I put this here as I understood that's where you put environment variables users are susceptible to edit for their needs. Now, this one is a bit special because if you wanted to put a different storage location (like a bucket on GCS or S3), you'd have to pass additional environment variables and thus modify the `docker-compose.yml` file anyway.

Is this usually how you are doing this to customize the behavior of some services?

File: docker-compose.yml
Comment (phacops) [reply docker-compose.yml]: I added `vroom` as a `depends_on` since `web` needs to POST to `vroom`, but then the consumers only depends on `kafka` technically. I think it's good like this.

**Review by hubertdeng123 — APPROVED** [review]

File: .env
Comment (BYK) [reply .env]: @phacops not really unless the setting applies to multiple services (like retention days). The typical way is to have a `<service_name>` folder with an example config we provide for defaults and then folks can customize it to their needs. See https://github.com/getsentry/self-hosted/tree/master/sentry or https://github.com/getsentry/self-hosted/tree/master/relay for examples.

**Comment by hubertdeng123** [comment]
23.5.2 is out now, this can be merged now

**Review by chadwhitacre — APPROVED** [review]

File: docker-compose.yml
Comment (aldy505) [inline docker-compose.yml:407]: The port shouldnt be exposed. 

```suggestion
```

File: docker-compose.yml
Comment (aldy505) [reply docker-compose.yml:432]: +1

File: docker-compose.yml
Comment (aldy505) [inline docker-compose.yml:404]: I'd prefer:

```suggestion
      SENTRY_BUCKET_PROFILES: "${SENTRY_BUCKET_PROFILES:-file://localhost//var/lib/sentry-profiles}"
```

That way, people can adjust their vroom storage backend config easier (ie. they can use GCS or S3 or Azure Blob), defined from .env or .env.custom without having them to change docker-compose.yml

Comment (aldy505) [reply docker-compose.yml:404]: Or, you might prefer `VROOM_BUCKET_PROFILES`.

**Review by aldy505 — COMMENTED** [review]
Just a little bit more..

Note about the exposed port: if you don't specify it as exposed port, that port will still be opened to other containers. Having a new exposed ports might cause a vulnerability for people who didn't set up a proper OS firewall.

File: docker-compose.yml
Comment (BYK) [reply docker-compose.yml:404]: I'd discourage this as it might be better to override the compose file with `docker-compose.override.yml` until we get a dedicated `vroom` config file.

Comment (aldy505) [reply docker-compose.yml:404]: Okay, agreed.

**Review by aldy505 — COMMENTED** [review]
LGTM!

File: docker-compose.yml
Comment (phacops) [reply docker-compose.yml:430]: I added to the comment instead, noting we don't need to persist this data but it's persisted anyway. I think removing it removes information about the kind of data stored.

File: scripts/bump-version.sh
Comment (chadwhitacre) [reply scripts/bump-version.sh:10]: https://github.com/getsentry/self-hosted/pull/2207 ftr, missing escape of the `|` led to [a corrupt 23.6.0 release](https://github.com/getsentry/self-hosted/blob/23.6.0/.env#L11-L12). 🐭

**Pull request merged**
